test/test_ali.py

```diff
@@ -44,180 +44,47 @@
-#  cut_json = egs_dir / "data/fbank/cuts_train-clean-360.json.gz"
-#  cut_json = egs_dir / "data/fbank/cuts_train-other-500.json.gz"
-#  ali_filename = ICEFALL_DIR / "egs/librispeech/ASR/data/ali_500/train-960.pt"
-
-cut_json = egs_dir / "data/fbank/cuts_test-clean.json.gz"
-ali_filename = ICEFALL_DIR / "egs/librispeech/ASR/data/ali_500/test_clean.pt"
+cuts_json = egs_dir / "data/token_ali/cuts_test-clean.json.gz"
 
 
 def data_exists():
-    return ali_filename.exists() and cut_json.exists() and lang_dir.exists()
+    return cuts_json.exists() and lang_dir.exists()
 
 
 def get_dataloader():
-    cuts_train = load_manifest(cut_json)
-    cuts_train = cuts_train.with_features_path_prefix(egs_dir)
-    train_sampler = SingleCutSampler(
-        cuts_train,
+    cuts = load_manifest(cuts_json)
+    cuts = cuts.with_features_path_prefix(egs_dir)
+    sampler = SingleCutSampler(
+        cuts,
         max_duration=40,
         shuffle=False,
     )
 
-    train = K2SpeechRecognitionDataset(return_cuts=True)
+    dataset = K2SpeechRecognitionDataset(return_cuts=True)
 
-    train_dl = DataLoader(
-        train,
-        sampler=train_sampler,
+    dl = DataLoader(
+        dataset,
+        sampler=sampler,
         batch_size=None,
         num_workers=1,
         persistent_workers=False,
     )
-    return train_dl
-
-
-def test_one_hot():
-    a = [1, 3, 2]
-    b = [1, 0, 4, 2]
-    c = [torch.tensor(a), torch.tensor(b)]
-    d = pad_sequence(c, batch_first=True, padding_value=0)
-    f = torch.nn.functional.one_hot(d, num_classes=5)
-    e = (1 - f) * -10.0
-    expected = torch.tensor(
-        [
-            [
-                [-10, 0, -10, -10, -10],
-                [-10, -10, -10, 0, -10],
-                [-10, -10, 0, -10, -10],
-                [0, -10, -10, -10, -10],
-            ],
-            [
-                [-10, 0, -10, -10, -10],
-                [0, -10, -10, -10, -10],
-                [-10, -10, -10, -10, 0],
-                [-10, -10, 0, -10, -10],
-            ],
-        ]
-    ).to(e.dtype)
-    assert torch.all(torch.eq(e, expected))
+    return dl
 
 
 def test():
-    """
-    The purpose of this test is to show that we can use pre-computed
-    alignments to construct a mask, adding it to a randomly generated
-    nnet_output, to decode the correct transcript from the resulting
-    nnet_output.
-    """
     if not data_exists():
         return
     device = torch.device("cpu")
     if torch.cuda.is_available():
         device = torch.device("cuda", 0)
     dl = get_dataloader()
-
-    subsampling_factor, ali = load_alignments(ali_filename)
-    ali = convert_alignments_to_tensor(ali, device=device)
-
-    lexicon = Lexicon(lang_dir)
-    max_token_id = max(lexicon.tokens)
-    num_classes = max_token_id + 1  # +1 for the blank
-    word_table = lexicon.word_table
-
-    HLG = k2.Fsa.from_dict(
-        torch.load(f"{lang_dir}/HLG.pt", map_location=device)
-    )
-
     for batch in dl:
```

Comment:
@pzelasko 
This line causes the following exception.

I think the interface for `K2SpeechRecognitionDataset` also needs to be updated to pass a `pad_value_dict`.

```
Traceback (most recent call last):
  File "test/test_ali.py", line 90, in <module>
    test()
  File "test/test_ali.py", line 82, in test
    for batch in dl:
  File "/ceph-fj/fangjun/py38/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 435, in __next__
    data = self._next_data()
  File "/ceph-fj/fangjun/py38/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1085, in _next_data
    return self._process_data(data)
  File "/ceph-fj/fangjun/py38/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1111, in _process_data
    data.reraise()
  File "/ceph-fj/fangjun/py38/lib/python3.8/site-packages/torch/_utils.py", line 428, in reraise
    raise self.exc_type(msg)
AssertionError: Caught AssertionError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/ceph-fj/fangjun/py38/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 198, in _worker_loop
    data = fetcher.fetch(index)
  File "/ceph-fj/fangjun/py38/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 46, in fetch
    data = self.dataset[possibly_batched_index]
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/dataset/speech_recognition.py", line 105, in __getitem__
    inputs, _ = self.input_strategy(cuts)
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/dataset/input_strategies.py", line 110, in __call__
    return collate_features(cuts, executor=_get_executor(self.num_workers))
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/dataset/collation.py", line 127, in collate_features
    cuts = maybe_pad(
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/dataset/collation.py", line 342, in maybe_pad
    return cuts.pad(
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/cut.py", line 3130, in pad
    return CutSet.from_cuts(
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/cut.py", line 2637, in from_cuts
    return CutSet(cuts=index_by_id_and_check(cuts))
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/utils.py", line 481, in index_by_id_and_check
    for m in manifests:
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/cut.py", line 3131, in <genexpr>
    cut.pad(
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/cut.py", line 1096, in pad
    return pad(
  File "/ceph-fj/fangjun/open-source-2/lhotse-ali/lhotse/cut.py", line 4137, in pad
    assert pad_value_dict is not None and all(
AssertionError: Cut being padded has custom TemporalArray attributes: ['token_alignment']. We expected a 'pad_value_dict' argument with padding values for these attributes.
```

Reply:
Gotcha. Yeah that assertion will be a pain in the butt. I'll change it to a warning and use a default value of 0 as you suggested.

Reply:
Thanks. No, it doesn't.

Reply:
One thing I don't quite understand about the exception is that the data loader does not read the custom fileds
contained in the CutSet, but why does it throw an exception related to custom fields?

Reply:
That's because the assertion was inside the `cut.pad` function; the assertion was supposed to be a sanity check to let you know that you forgot to specify the pad value for custom fields (if any are present). I only realized later that in many use-cases you actually won't care about the custom attributes presence so it makes sense to use a default value instead.

Reply:
(the K2 dataset is padding mini-batches and that triggered the assertion)